sway-error/src/error.rs

```diff
@@ -1818,6 +1831,62 @@ impl ToDiagnostic for CompileError {
                 },
                 help: vec![],
             },
+            TupleIndexOutOfBounds { index, count, tuple_type, referencing_level, span, prefix_span } => Diagnostic {
+                reason: Some(Reason::new(code(1), "Tuple index is out of bounds".to_string())),
+                issue: Issue::error(
+                    source_engine,
+                    span.clone(),
+                    format!("Tuple index {index} is out of bounds. The {}tuple has only {count} element{}.",
+                        if *referencing_level > 0 {
+                            "referenced "
+                        } else {
+                            ""
+                        },
+                        plural_s(*count)
+                    )
+                ),
+                hints: vec![
+                    Hint::info(
+                        source_engine,
+                        prefix_span.clone(),
+                        format!("This expression has type \"{tuple_type}\".")
+                    ),
+                    Hint::info(
+                        source_engine,
+                        prefix_span.clone(),
+                        format!("That is {} tuple of only {} element{}.",
+                            match *referencing_level {
+                                0 => "a",
+                                1 => "a reference to a",
+                                _ => "an indirect reference to a",
```

Comment:
I think having this whole infrastructure and concept of a reference depth just to have the right turn of phrase is overengineering.

Devs will expect the dereferencing to work through refs, we shouldn't have to explain it, and if we do we should not do it in terms this specific.

I think this should just say "The tuple has only x elements", provide the type and ditch this last hint along with all references to a referencing level.